### PR TITLE
Add active SVM feature

### PR DESCRIPTION
**Pull request by m-chaturvedi**

### CODE REVIEW

**Comment by Purg** [comment]
Remember that the libsvm_hik class is extending an interface and this update now causes the `LibSvmHikRelevancyIndex` implementation to deviate from this interface.  Now, there is a valid point here of whether the parent interface should be updated/modified.  I'm not sure the description of the `feedback_pool`, where descriptors are mapped to their distance from a separating plane, is applicable outside of SVM, or similar, implementations. I.e. if we made a deep-learning implementation of the `RelevancyIndex` interface, would that also have a notion of a separating plane that can be measured against?

File: python/smqtk/algorithms/relevancy_index/libsvm_hik.py [hunk 1]
```diff
@@ -189,6 +190,49 @@ def get_vector(d):
             with open(self.descr_cache_fp, 'wb') as f:
                 cPickle.dump(self._descr_cache, f, -1)
 
+    def get_descr_and_distance_list(self, svm_model):
```
Comment (Purg) [inline python/smqtk/algorithms/relevancy_index/libsvm_hik.py]: Missing description and parameter/return documentation.

File: python/smqtk/algorithms/relevancy_index/libsvm_hik.py [hunk 2]
```diff
@@ -189,6 +190,49 @@ def get_vector(d):
...
+            dict_self_descr_cache.append(dict(zip(range(len(d.vector())),
+                                                  list(d.vector()))))
+            svm_node_array, _ = svm.gen_svm_nodearray(list(d.vector()))
+            distance = svmutil.svm_distance_from_plane(svm_node_array,
+                                                       svm_model)
+            d_for_unlabelled.append(abs(distance[0]))
+
+        return dict_self_descr_cache, d_for_unlabelled
+
+    def get_rank_and_feedback_using_libsvm(self, svm_model):
```
Comment (Purg) [inline python/smqtk/algorithms/relevancy_index/libsvm_hik.py]: Missing description and parameter/return documentation.

File: python/smqtk/algorithms/relevancy_index/libsvm_hik.py [hunk 3]
```diff
@@ -202,9 +246,15 @@ def rank(self, pos, neg):
...
+        :return: A tuple consisting of rank_pool and feedback_pool, where the
+            rank pool is a map of indexed descriptor elements to a rank value
+            between [0, 1] (inclusive) range, where a 1.0 means most relevant
+            and 0.0 meaning least relevant.
+            And the feedback_pool is a an ordered map mapping descriptor with
+            their distance from plane. The order corresponds to the increasing 
+            order of distance of the descriptors from the separating plane.
+
+
         :rtype: dict[smqtk.representation.DescriptorElement, float]
```
Comment (Purg) [inline python/smqtk/algorithms/relevancy_index/libsvm_hik.py]: Type documentation not updated for new return format.

File: python/smqtk/iqr/iqr_session.py [hunk 4]
```diff
@@ -171,6 +174,20 @@ def ordered_results(self):
                                     reverse=True))
             return None
 
+    def ordered_feedback(self):
+        """
+        Return a tuple of (id, distance) in order of increasing distance from
+        the separating plane for SSVM.
```
Comment (Purg) [inline python/smqtk/iqr/iqr_session.py]: Is ``SSVM`` intentional, or should this be `SVM`?

**Comment by m-chaturvedi** [comment]
My guess would be no, this separating plane distance should be SVM related, but my deep-learning knowledge is rather limited.  However, those distances, are something like uncertainty measures in some sense.  We want to get the input of the user on points which we are the most uncertain about.  

There must be other ways of getting IQR with deep learning.  So, a `feedback_pool` variable may make sense in general.  It being an ordered map, ordered in decreasing order of importance of queries.

**Comment by Purg** [comment]
I think you hit a good description in that we're trying to describe the uncertainty with the `feedback_pool`'s values where I'm assuming lower values (closer to hyperplane) mean less certainty.  If we want to change the parent interface, then the return type should be something different from a dictionary mapping DElem->float.

**Comment by Purg** [comment]
Given that this is for viame/master, I'm not sure what @mattdawkins priorities are.  Interface work may need to be done separately on master.

**Pull request closed without merging**